iml-agent/src/daemon_plugins/ntp.rs

```diff
@@ -0,0 +1,536 @@
+polling server every 64 s"#
+                .into(),
+        )
+    }
+
+    fn get_chronyc_ntpdata_command() -> impl Future<Item = String, Error = ImlAgentError> {
+        future::ok::<String, ImlAgentError>(
+            r#"
+
+Remote address  : 10.73.10.10 (0A490A0A)
+Remote port     : 123
+Local address   : 10.73.10.12 (0A490A0C)
+Leap status     : Normal
+Version         : 4
+Mode            : Server
+Stratum         : 11
+Poll interval   : 4 (16 seconds)
+Precision       : -25 (0.000000030 seconds)
+Root delay      : 0.000000 seconds
+Root dispersion : 0.011124 seconds
+Reference ID    : 7F7F0100 ()
+Reference time  : Fri Sep 13 08:03:04 2019
+Offset          : +0.000026767 seconds
+Peer delay      : 0.000403893 seconds
+Peer dispersion : 0.000000056 seconds
+Response time   : 0.000079196 seconds
+Jitter asymmetry: +0.00
+NTP tests       : 111 111 1111
+Interleaved     : No
+Authenticated   : No
+TX timestamping : Kernel
+RX timestamping : Kernel
+Total TX        : 1129
+Total RX        : 1129
+Total valid RX  : 1129"#
+                .into(),
+        )
+    }
+
+    fn is_ntp_synced_command() -> impl Future<Item = String, Error = ImlAgentError> {
+        future::ok::<String, ImlAgentError>("b true".into())
+    }
+
+    #[test]
+    fn test_is_ntp_synced() {
+        let s: String = r#"b true"#.into();
+
+        assert_eq!(is_ntp_synced(s), TimeSync::Synced);
+    }
+
+    #[test]
+    fn test_is_ntp_not_synced() {
+        let s: String = r#"b false"#.into();
+
+        assert_eq!(is_ntp_synced(s), TimeSync::Unsynced);
+    }
+
+    #[test]
+    fn test_get_ntp_time_offset() {
+        let s: String = r#"synchronised to NTP server (10.73.10.10) at stratum 12
+time correct to within 949 ms
+polling server every 64 s"#
+            .into();
+
+        assert_eq!(get_ntp_time_offset(s), Some("949 ms".into()));
+    }
+
+    #[test]
+    fn test_get_chrony_time_offset() {
+        let s: String = r#"
+
+Remote address  : 10.73.10.10 (0A490A0A)
+Remote port     : 123
+Local address   : 10.73.10.12 (0A490A0C)
+Leap status     : Normal
+Version         : 4
+Mode            : Server
+Stratum         : 11
+Poll interval   : 4 (16 seconds)
+Precision       : -25 (0.000000030 seconds)
+Root delay      : 0.000000 seconds
+Root dispersion : 0.011124 seconds
+Reference ID    : 7F7F0100 ()
+Reference time  : Fri Sep 13 08:03:04 2019
+Offset          : +0.000026767 seconds
+Peer delay      : 0.000403893 seconds
+Peer dispersion : 0.000000056 seconds
+Response time   : 0.000079196 seconds
+Jitter asymmetry: +0.00
+NTP tests       : 111 111 1111
+Interleaved     : No
+Authenticated   : No
+TX timestamping : Kernel
+RX timestamping : Kernel
+Total TX        : 1129
+Total RX        : 1129
+Total valid RX  : 1129"#
+            .into();
+
+        assert_eq!(
+            get_chrony_time_offset(s),
+            Some("0.000026767 seconds".into())
+        );
+    }
+
+    #[test]
+    fn test_session_with_ntp_configured_by_iml() {
```

Comment:
I'm trying to find a better way to do this. The idea is that I implement the SideEffects trait for the TestNtp struct. This works but unfortunately I have to re-implement the DaemonPlugin trait. I was hoping I could somehow wrap the Ntp struct, implement the SideEffect trait, and then I wouldn't need to re-implement DaemonPlugin. This seems to be more difficult than I anticipated though.

Reply:
Why do we need to implement an entire `SideEffects` trait? I.E. What do we gain by testing DaemonPlugin instead of testing individual calls?